rclcpp/src/rclcpp/clock.cpp

```diff
@@ -182,6 +182,75 @@ Clock::sleep_for(Duration rel_time, Context::SharedPtr context)
   return sleep_until(now() + rel_time, context);
 }
 
+bool
+Clock::is_valid()
+{
+  // Checks for null pointer, missing get_now() implementation, and RCL_CLOCK_UNINITIALIZED
+  if (!rcl_clock_valid(get_clock_handle())) {
+    RCUTILS_LOG_WARN("clock is uninitialized and will never become valid");
+    return false;
+  }
+
+  switch (get_clock_type()) {
+    case RCL_ROS_TIME:
+    case RCL_STEADY_TIME:
+    case RCL_SYSTEM_TIME:
+      return now().nanoseconds() > 0;
+
+    // By right we shouldn't even get to this block, but these cases are included for completeness
+    case RCL_CLOCK_UNINITIALIZED:
+    default:
+      return false;
+  }
+}
+
+bool
+Clock::wait_for_valid(Context::SharedPtr context)
+{
+  if (!context || !context->is_valid()) {
+    throw std::runtime_error("context cannot be slept with because it's invalid");
+  }
+
+  if (!rcl_clock_valid(get_clock_handle())) {
+    throw std::runtime_error("clock cannot be waited on as it is not rcl_clock_valid");
+  }
+
+  if (is_valid()) {
+    return true;
+  } else {
+    // Wait until the first valid time
+    return sleep_until(rclcpp::Time(0, 1, get_clock_type()), context);
+  }
+}
+
+bool
+Clock::wait_for_valid(
+  const Duration & timeout,
+  Context::SharedPtr context,
+  Duration wait_tick_ns)
+{
+  if (!context || !context->is_valid()) {
+    throw std::runtime_error("context cannot be slept with because it's invalid");
+  }
+
+  if (!rcl_clock_valid(get_clock_handle())) {
+    throw std::runtime_error("clock cannot be waited on as it is not rcl_clock_valid");
+  }
+
+  Clock timeout_clock = Clock(RCL_STEADY_TIME);
+  Time start = timeout_clock.now();
+
+  while (!is_valid() && context->is_valid()) {  // Context check checks for rclcpp::shutdown()
+    timeout_clock.sleep_for(Duration(wait_tick_ns));
```

Comment:
if `Duration(wait_tick_ns)` >> timeout, it will not be timed out with specified timeout Duration?
i think we need to add `if` statement to check if `wait_tick_ns` is less than `timeout`, and calculate for next sleep time based on how much time left until timeout.
test would be ideal.

Reply:
> But this is a good question; it would be nice if this policy were encoded in the `rcl` layer, and then just inherited by the client libraries. @methylDragon Do you think that is feasible?

It'll be nice to put this in the `rcl` layer, though I'm not too sure how to do it (well, more precisely, how to name the methods without causing API/ABI breaks...)
We need a distinct way to check if the members of a clock are initialized (currently what `rcl_clock_valid()` is doing), since that gets used by the client libraries (e.g. in the `ros_time_active()` method, which then feeds into the clock setting methods).

So if I were to add it into the `rcl` layer, I need to add a new method. The problem is:
- I either need to change the current `rcl_clock_valid()` and then have this new method be the `rcl_clock_valid()`, necessitating changes across every client library that uses the method. But I'm pretty sure this will break ABI/API.
- Or I need to add a new method, but then it'd be confusing because `rcl_clock_valid()` would not be encoding the notion of validity in the design docs... Is there a better method name for the new method I can use?

If we can't resolve the naming issue, I think it'd be more intuitive to place it in the client libraries? Or maybe I'm missing something :V

Reply:
https://github.com/ros2/rclcpp/pull/2040/commits/6face35ae9b23405eb705169d767197f9101446c